src/shared/calendarspec.c

```diff
@@ -1251,11 +1276,13 @@ static int find_next(const CalendarSpec *spec, struct tm *tm, usec_t *usec) {
 
         c = *tm;
         tm_usec = *usec;
+        dst = spec->dst;
 
         for (unsigned iteration = 0; iteration < MAX_CALENDAR_ITERATIONS; iteration++) {
                 /* Normalize the current date */
                 (void) mktime_or_timegm_usec(&c, spec->utc, /* ret= */ NULL);
-                c.tm_isdst = spec->dst;
+                if (dst != -2)
+                        c.tm_isdst = dst;
```

Comment:
hmm we should unset the flag here again, no? i.e. `invalidate_dst = false;` here, to make sure we only revalidate the thing once after each check

Reply:
Sorry, I don't think I understand your idea - if we turn `dst` into a bool then we have two sources of `tm_isdst` - `spec->dst` and `c.tm_isdst` (which is set by mktime on each iteration before being overwritten). The `invalidate_dst` flag would determine which source would we use - first `spec->dst` is used but if we are stuck, we switch to mktime's `tm_isdst`.
Once the flag is flipped, we cannot update the spec's `dst` as the caller expects us to get a const pointer.

Reply:
i am not sure i follow? you only need the -2 thing locally, so you can replace that with a local variable no? that simply controls which source to use?

Reply:
The "make sure we only revalidate the thing once after each check" confused me.
I have replaced `-2` with a local variable as you suggested - if true then keep the value of `c.tm_isdst` that was stored by `mktime`, otherwise overwrite (as the function did before) it with the value of `spec->dst`.